### PR TITLE
refactor(migrations): support use of an ESM `@angular/compiler-cli` package

### PR DESCRIPTION
**Pull request by clydin**
Currently, migrations and schematics must be in CommonJS format. However, framework packages will only be ESM from v13 and onward. To support this configuration, dynamic import expressions are now used to load `@angular/compiler-cli` and its new secondary entry point `@angular/compiler-cli/private/migrations`. Dynamic imports within Node.js allow the `@angular/core` migrations’ CommonJS code to load ESM code. Unfortunately, TypeScript will currently, unconditionally down-level dynamic import into a require call. `require` calls cannot load ESM code and will result in a runtime error. To workaround this, a Function constructor is used to prevent TypeScript from changing the dynamic import. Once TypeScript provides support for keeping the dynamic import this workaround can be dropped and replaced with a standard dynamic import.  Due to the use of the dynamic import, a reference to the dynamically loaded modules must now be passed to all locations that use values from the modules.

### CODE REVIEW

**Review by devversion — APPROVED** [review]
LGTM. Hopefully we can use ESM for the migrations in the future as this would simplify things a lot.

File: packages/core/schematics/utils/load_esm.ts [hunk 1]
```diff
@@ -69,15 +41,15 @@ export async function loadEsmModule<T>(modulePath: string|URL): Promise<T> {
...
-    // above does not exist in which case `@angular/compiler-cli` will still be CommonJS.
-    // TODO_ESM: This can be removed once the `@angular/compiler-cli/private/migrations` entry
-    // point exists.
+    // rules_nodejs currently does not support exports field entries. This is a temporary workaround
+    // that leverages devmode currently being CommonJS. If that changes before rules_nodejs supports
+    // exports then this workaround needs to be reworked.
+    // TODO_ESM: This can be removed once Bazel supports exports fields.
     return {
       ...require('@angular/compiler-cli/src/ngtsc/annotations'),
       ...require('@angular/compiler-cli/src/ngtsc/imports'),
```
Comment (devversion) [inline packages/core/schematics/utils/load_esm.ts:55]: I think the fallback should could just be: `require('@angular/compiler-cli/src/ngtsc/migrations.js);` (notice the explicit extension)

**Review by devversion — APPROVED** [review]
Still LGTM!

**Review by alxhub — APPROVED** [review]

**Comment by alxhub** [comment]
Caretaker: cl/400826396 is a necessary update to the g3 patch for this change. Please coordinate with me to land these together.

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit 6d1c2f79da9e034308c7f2cfbd16499e0cfd51f9.

**Pull request closed without merging**